### PR TITLE
kvs: implement lightweight key watch and API

### PR DESCRIPTION
**Pull request by garlick**
This PR finishes the alternate KVS watch API, based on passing a FLUX_KVS_WATCH flag to `flux_kvs_lookup()`; and the lightweight `kvs-watch` service, which only checks watched keys when they are mentioned in a commit.   This strategy was discussed in #1556.

N.B. this PR falls short of implementing the "bulk" watch described in #1631 - this is for another PR.

In addition to the API and `kvs-watch` modifications, the `flux kvs get` subcommand now has a `--watch` and related options.

Still TODO:  doc updates and sharness test(s).

### CODE REVIEW

**Comment by chu11** [comment]
looked through with a quick skim and nothing looked too out of place.  Will wait for tests and take a deeper look at that time.

**Comment by garlick** [comment]
Thanks!  I added a couple which greatly helps coverage but there are a bunch of corner cases to check still.

**Comment by garlick** [comment]
This is probably getting close and could use some review by another pair of eyes.

File: src/cmd/flux-kvs.c [hunk 1]
```diff
@@ -96,6 +96,18 @@ static struct optparse_option get_opts[] =  {
     { .name = "at", .key = 'a', .has_arg = 1,
       .usage = "Lookup relative to RFC 11 snapshot reference",
     },
+    { .name = "order", .key = 'o', .has_arg = 1,
```
Comment (chu11) [inline src/cmd/flux-kvs.c]: I don't think ```has_arg``` needs to be 1?

File: t/t1007-kvs-lookup-watch.t [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+test_expect_success 'flux kvs get --watch allows owner access to guest ns' '
+	flux kvs namespace-create --owner=9999 testns5 &&
+	flux kvs --namespace=testns5 put test.k=100 &&
+	! FLUX_HANDLE_ROLEMASK=0x1 FLUX_HANDLE_USERID=9998 \
+		flux kvs --namespace=testns4 get --watch --count=1 test.k &&
+	flux kvs namespace-remove testns4
+'
+
+test_done
```
Comment (chu11) [inline t/t1007-kvs-lookup-watch.t]: not sure if I'm missing it here on the github review, but is ```-l``` or ```-o``` tested?

File: src/common/libflux/rpc.h [hunk 3]
```diff
@@ -22,6 +22,9 @@ flux_future_t *flux_rpc_raw (flux_t *h, const char *topic,
                              const void *data, int len,
                              uint32_t nodeid, int flags);
 
+flux_future_t *flux_rpc_message (flux_t *h, const flux_msg_t *msg,
```
Comment (chu11) [inline src/common/libflux/rpc.h]: do we need a new manpage entry for this function?

File: src/modules/kvs-watch/kvs-watch.c [hunk 4]
```diff
@@ -176,30 +205,139 @@ static struct namespace *namespace_create (struct watch_ctx *ctx,
...
+    json_decref (o);
+    flux_msg_destroy (msg);
+    errno = saved_errno;
+    return NULL;
+}
+
 /* Respond to watcher request, if appropriate.
  * De-list and destroy watcher from namespace on error.
- * De-hash and destroy namespace if watchers list becomes empty.
+ * De-hash and destroy namespace if watchers list become empty.
```
Comment (chu11) [inline src/modules/kvs-watch/kvs-watch.c]: i actually think this should be ```becomes```? :-)

File: src/modules/kvs-watch/kvs-watch.c [hunk 5]
```diff
@@ -229,6 +371,42 @@ static void watcher_respond (struct namespace *ns, struct watcher *w)
         }
         w->rootseq = ns->commit->rootseq;
     }
+    /* flux_kvs_lookup (FLUX_KVS_WATCH)
+     *
+     * Ordering note: KVS lookups can be returned out of order, so lookup
+     * futures are added to the w->lookups zlist in commit order, and an
+     * out of order result is queued until all the lookups ahead of it have
+     * been retired.
+     *
```
Comment (chu11) [inline src/modules/kvs-watch/kvs-watch.c]: This comment sort of confused me.  I think it's b/c the latter half of what this comment says is implemented in the ```lookup_continuation``` function.  Perhaps it should be split up a bit?

File: src/cmd/flux-kvs.c (hunk 1 above)
Comment (grondo) [reply src/cmd/flux-kvs.c]: Also, just a minor suggestion to use `-s, --sort` since many other tools do so

Comment (chu11) [reply src/cmd/flux-kvs.c]: I'm not sure ```sort``` is the right word.  What we're looking for is a word for "output one key at a time".  Perhaps ```order``` is not the best, but not sure what would be better.

File: src/cmd/flux-kvs.c [hunk 6]
```diff
@@ -228,7 +240,7 @@ static struct optparse_subcommand subcommands[] = {
       NULL
     },
     { "get",
-      "[-j|-r|-t] [-a treeobj] key [key...]",
+      "[-j|-r|-t] [-a treeobj] [-o] [-l] [-w] [-c COUNT] key [key...]",
```
Comment (chu11) [inline src/cmd/flux-kvs.c]: just realized, need to add to ```flux-kvs.adoc``` too.

**Comment by garlick** [comment]
>One high level question. Was the intent to eventually remove the original "KVS watch" in a later PR?

Yes but this PR is insufficient, due to the semantic differences pointed out in #1556 (including the one you mentioned).  I think we'll need something like

FLUX_KVS_WATCH_CLASSIC = (FLUX_KVS_WATCH | FLUX_KVS_WAITCREATE | FLUX_KVS_PEDANTIC)

where FLUX_KVS_WAITCREATE is described in #1651 

FLUX_KVS_PEDANTIC (or a better name) would check the key on every commit to its namespace (like current watch).  I mean to open an issue on this flag as well.

The problem of returning multiple values could fall out of PEDANTIC, or it could be made part of the functionality on this PR.  I could go either way - thoughts?

File: src/cmd/flux-kvs.c (hunk 1 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: Oops - thanks for pointing out the has_arg problem.

The `--order` option ensures that values are printed in the same order as specified on the command line, as opposed to random order if they are allowed to be fetched in parallel.   Maybe `--serial-lookup` would be a better option name?

File: t/t1007-kvs-lookup-watch.t (hunk 2 above)
Comment (garlick) [reply t/t1007-kvs-lookup-watch.t]: True, I think I should add some basic tests for those options.

File: src/common/libflux/rpc.h (hunk 3 above)
Comment (garlick) [reply src/common/libflux/rpc.h]: yes thanks!

File: src/cmd/flux-kvs.c (hunk 1 above)
Comment (grondo) [reply src/cmd/flux-kvs.c]: Ah, sorry! I was confused by the usage message
```
      .usage = "For multiple keys, ensure values are output in order",
```

I still don't understand what the option does exactly, but you are definitely right `--sort` isn't correct. Sorry for the noise. (On another note, did the `flux-kvs(1)` manpage need an update here?

File: src/modules/kvs-watch/kvs-watch.c (hunk 4 above)
Comment (garlick) [reply src/modules/kvs-watch/kvs-watch.c]: yep.

File: src/cmd/flux-kvs.c (hunk 6 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: Oh yeah!  Thanks.

**Comment by grondo** [comment]
I'm just curious -- what happens if a user destroys a future with FLUX_KVS_WATCH without first calling `flux_kvs_lookup_cancel`?  Will responses continue to be sent, but dropped? I see that disconnects also trigger a cancellation in the kvs-watch module, but will this cancel generate spurious ENODATA messages to a handle we know has disconnected? Sorry for the trivial questions.

**Comment by garlick** [comment]
> what happens if a user destroys a future with FLUX_KVS_WATCH without first calling flux_kvs_lookup_cancel? Will responses continue to be sent, but dropped?

Yes.  They will be dropped if the future was reset and then destroyed without cancelling the watch, because the matchtag will be "leaked" by the RPC code.  If the code is really wrong and the future is destroyed in a fulfilled state, the matchtag could be recycled, and erstwhile watch responses could be mixed up with other responses.

> I see that disconnects also trigger a cancellation in the kvs-watch module, but will this cancel generate spurious ENODATA messages to a handle we know has disconnected?

The `w->mute` flag should prevent a response if cancelled due to disconnect.

File: src/cmd/flux-kvs.c (hunk 1 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: The code was changed to be asynchronous, so without the option, keys could be output in any order.   Maybe I should just make this the default if `--watch` is not specified and we can avoid this option entirely?

Comment (garlick) [reply src/cmd/flux-kvs.c]: I think that's what I should do.  Shaving off a few milliseconds for a multiple-key get that is rarely used is not worth much, while making the command confusing does harm.

**Comment by grondo** [comment]
Great, thanks for the answers!

File: src/cmd/flux-kvs.c (hunk 1 above)
Comment (grondo) [reply src/cmd/flux-kvs.c]: Sounds good to me, thanks!

**Comment by garlick** [comment]
I think I've addressed the comments so far.  LMK if I should squash and rebase.

**Comment by chu11** [comment]
Looks good to me to squash and merge.  One English nit, which I'm even debating if it should be changed.  But I'll mention it:

```
If '-w', after the initial value, display the new value each time the key changes until interrupted,
```

Technically this isn't correct b/c writing the same value will lead to watch outputs.  But, 99% of the time this is what happens and it is how people should think of the option working.  So I think it's ok to leave the text like this?

**Comment by garlick** [comment]
I don't think we need to get to that level of detail in the `flux kvs` man page.

Possibly when we implement #1653 it will turn out to be easiest to just do the value check unconditionally anyway.  Will have to see what it adds to "weight" of the operation.  Maybe not much.

**Comment by chu11** [comment]
lgtm, will merge once travis passes

**Comment by garlick** [comment]
Had to restart two stalled builders and one (apparently) spurious valgrind error, but it finally passed.

**Comment by garlick** [comment]
Thanks!
